src/Bicep.Local.Deploy/Extensibility/LocalExtensionFactoryManager.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Bicep.Core.TypeSystem.Types;
+using Bicep.Local.Extension.Protocol;
+using Google.Protobuf;
+
+namespace Bicep.Local.Deploy.Extensibility
+{
+    public record LocalExtensionKey(string ExtensionName, string ExtensionVersion);
+
+    public class LocalExtensionFactoryManager(ILocalExtensionFactory extensionFactory) : ILocalExtensionFactoryManager
+    {
+        private readonly Dictionary<LocalExtensionKey, Func<Task<ILocalExtension>>> RegisteredLocalExtensions = [];
+        private readonly Dictionary<LocalExtensionKey, ILocalExtension> InitializedLocalExtensions = [];
```

Comment:
Worst case, space complexity will be O(2n), which equates to O(n), with this new struct I'll be able to load extensions lazily - without using `Lazy` or `AsyncLazy` - using AsyncLazy might be a better approach, I just wasn't able to get it to work.